cmd/cvp-trigger/main.go

```diff
@@ -0,0 +1,149 @@
+
+	"github.com/openshift/ci-tools/pkg/rehearse"
+	"github.com/openshift/ci-tools/pkg/util"
+)
+
+const (
+	prowConfigPathOption = "prow-config-path"
+	jobConfigPathOption  = "job-config-path"
+	periodicOption       = "periodic"
+)
+
+type options struct {
+	dry bool
+
+	prowConfigPath string
+	jobConfigPath  string
+
+	periodic string
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.BoolVar(&o.dry, "dry-run", true, "Whether to actually submit the job to Prow")
+	fs.StringVar(&o.prowConfigPath, prowConfigPathOption, "", "Path to the Prow config file")
+	fs.StringVar(&o.jobConfigPath, jobConfigPathOption, "", "Path to the Prow job config directory")
+	fs.StringVar(&o.periodic, "periodic", "", "Name of the Periodic job to manually trigger")
+
+	_ = fs.Parse(os.Args[1:])
+	return o
+}
+
+func (o options) validate() error {
+	if o.prowConfigPath == "" {
+		return fmt.Errorf("required parameter %s was not provided", prowConfigPathOption)
+	}
+
+	if o.jobConfigPath == "" {
+		return fmt.Errorf("required parameter %s was not provided", jobConfigPathOption)
+	}
+
+	if o.periodic == "" {
+		return fmt.Errorf("required parameter %s was not provided", periodicOption)
+	}
+
+	return nil
+}
+
+func main() {
+	o := gatherOptions()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("incorrect options")
+	}
+
+	go func() {
+		interrupts.WaitForGracefulShutdown()
+		os.Exit(1)
+	}()
+
+	config, err := prowconfig.Load(o.prowConfigPath, o.jobConfigPath)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to read Prow configuration")
+	}
+
+	var selected *prowconfig.Periodic
+	for _, job := range config.AllPeriodics() {
+		if job.Name == o.periodic {
+			selected = &job
+			break
+		}
+	}
+
+	if selected == nil {
+		logrus.WithField("job-name", o.periodic).Fatal("failed to find the job")
+	}
+
+	var clusterConfig *rest.Config
+	if !o.dry {
+		clusterConfig, err = util.LoadClusterConfig()
+		if err != nil {
+			logrus.WithError(err).Fatal("failed to load cluster configuration")
+		}
+	}
+
+	pjclient, err := rehearse.NewProwJobClient(clusterConfig, config.ProwJobNamespace, o.dry)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to create a ProwJob client")
+	}
+
+	prowjob := pjutil.NewProwJob(pjutil.PeriodicSpec(*selected), nil, nil)
+	logrus.WithFields(pjutil.ProwJobFields(&prowjob)).Info("submitting a new prowjob")
+	created, err := pjclient.Create(&prowjob)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to submit the prowjob")
+	}
+
+	if o.dry {
+		jobAsYAML, err := yaml.Marshal(prowjob)
+		if err != nil {
+			logrus.WithError(err).Fatal("failed to marshal the prowjob to YAML")
+		}
+		fmt.Printf(string(jobAsYAML))
+		os.Exit(0)
+	}
+
+	logger := logrus.WithFields(pjutil.ProwJobFields(created))
+	logger.Info("submitted the prowjob, waiting for its result")
+
+	selector := fields.SelectorFromSet(map[string]string{"metadata.name": created.Name})
+
+	for {
```

Comment:
If you exit on error from the Watch why do you need this outer `for`?

Reply:
I followed what the rehearsals are doing, and I vaguely recollect the reason why we did that was that sometimes the result channel closed prematurely.